sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala

```diff
@@ -0,0 +1,255 @@
+    }
+
+    loadTable(ident)
+  }
+
+  override def alterTable(
+      ident: Identifier,
+      changes: TableChange*): Table = {
+    val catalogTable = try {
+      catalog.getTableMetadata(ident.asTableIdentifier)
+    } catch {
+      case _: NoSuchTableException =>
+        throw new NoSuchTableException(ident)
+    }
+
+    val properties = CatalogV2Util.applyPropertiesChanges(catalogTable.properties, changes)
+    val schema = CatalogV2Util.applySchemaChanges(catalogTable.schema, changes)
+
+    try {
+      catalog.alterTable(catalogTable.copy(properties = properties, schema = schema))
+    } catch {
+      case _: NoSuchTableException =>
+        throw new NoSuchTableException(ident)
+    }
+
+    loadTable(ident)
+  }
+
+  override def dropTable(ident: Identifier): Boolean = {
+    try {
+      if (loadTable(ident) != null) {
+        catalog.dropTable(
+          ident.asTableIdentifier,
+          ignoreIfNotExists = true,
+          purge = true /* skip HDFS trash */)
+        true
+      } else {
+        false
+      }
+    } catch {
+      case _: NoSuchTableException =>
+        false
+    }
+  }
+
+  implicit class TableIdentifierHelper(ident: Identifier) {
+    def asTableIdentifier: TableIdentifier = {
+      ident.namespace match {
+        case Array(db) =>
+          TableIdentifier(ident.name, Some(db))
+        case Array() =>
+          TableIdentifier(ident.name, Some(catalog.getCurrentDatabase))
+        case _ =>
+          throw new NoSuchTableException(ident)
+      }
+    }
+  }
+
+  override def toString: String = s"V2SessionCatalog($name)"
+}
+
+/**
+ * An implementation of catalog v2 [[Table]] to expose v1 table metadata.
+ */
+case class CatalogTableAsV2(v1Table: CatalogTable) extends Table {
+  implicit class IdentifierHelper(identifier: TableIdentifier) {
+    def quoted: String = {
+      identifier.database match {
+        case Some(db) =>
+          Seq(db, identifier.table).map(quote).mkString(".")
+        case _ =>
+          quote(identifier.table)
+
+      }
+    }
+
+    private def quote(part: String): String = {
+      if (part.contains(".") || part.contains("`")) {
+        s"`${part.replace("`", "``")}`"
+      } else {
+        part
+      }
+    }
+  }
+
+  def catalogTable: CatalogTable = v1Table
+
+  lazy val options: Map[String, String] = {
+    v1Table.storage.locationUri match {
+      case Some(uri) =>
+        v1Table.storage.properties + ("path" -> uri.toString)
+      case _ =>
+        v1Table.storage.properties
+    }
+  }
+
+  override lazy val properties: util.Map[String, String] = v1Table.properties.asJava
+
+  override lazy val schema: StructType = v1Table.schema
+
+  override lazy val partitioning: Array[Transform] = {
+    val partitions = new mutable.ArrayBuffer[Transform]()
+
+    v1Table.partitionColumnNames.foreach { col =>
+      partitions += LogicalExpressions.identity(col)
+    }
+
+    v1Table.bucketSpec.foreach { spec =>
```

Comment:
just out of curiosity, does the ordering here matter? Do partitions need to be specified as Transforms first, and then bucketing is specified? Could bucketing be specified first?

Reply:
Bucketing can be specified first by using `bucket(16, col)` as a transform. When it is created using the `BUCKET BY` clause, we add it last because that's where it is effectively added in Hive and Spark implementations. But the new syntax allows users to specify the order that they choose.